### PR TITLE
[wasm-split] Read and use profiles

### PR DESCRIPTION
**Pull request by tlively**
Read the profiles produced by wasm-split's instrumentation to guide splitting.
In this initial implementation, all functions that the profile shows to have
been called are kept in the initial module. In the future, users may be able to
tune this so that functions that are run later will still be split out.

cc @sbc100 @awtcode

### CODE REVIEW

File: src/tools/wasm-split.cpp [hunk 1]
```diff
@@ -443,12 +446,50 @@ void instrumentModule(Module& wasm, const WasmSplitOptions& options) {
   writer.write(wasm, options.output);
 }
 
+std::set<Name> readProfile(Module& wasm, const WasmSplitOptions& options) {
```
Comment (kripken) [inline src/tools/wasm-split.cpp:449]: where is the profile file documented? I'm unsure what to compare the code here against.

File: test/lit/wasm-split/call_exports.mjs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// Instantiates an instrumented module, calls the given exports, then collects
+// its wasm-split profile and writes it to a given file.
+//
+// Usage:
+//
+//  node call_exports.mjs <module> <profile> <export>*
```
Comment (kripken) [inline test/lit/wasm-split/call_exports.mjs:6]: what does this lit test do? I don't see any checks for comparing anything... maybe I just don't know enough about lit.

File: test/lit/wasm-split/profile-guided.wast [hunk 3]
```diff
@@ -0,0 +1,67 @@
+;; Instrument the binary
+
+;; RUN: wasm-split --instrument %s -o %t.instrumented.wasm
```
Comment (kripken) [inline test/lit/wasm-split/profile-guided.wast:3]: what does such a line do without a check later? just verify it doesn't crash?

File: test/lit/wasm-split/profile-guided.wast [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+;; RUN: node %S/call_exports.mjs %t.instrumented.wasm %t.bar.prof bar
+;; RUN: node %S/call_exports.mjs %t.instrumented.wasm %t.both.prof foo bar
+;; RUN: node %S/call_exports.mjs %t.instrumented.wasm %t.none.prof
+
+;; Create profile-guided splits
+
+;; RUN: wasm-split %s --profile=%t.foo.prof -v -o1 %t.foo.1.wasm -o2 %t.foo.2.wasm \
+;; RUN:   | filecheck %s --check-prefix FOO
+
+;; FOO: Keeping functions: deep_foo_callee, foo, foo_callee, shared_callee
```
Comment (kripken) [inline test/lit/wasm-split/profile-guided.wast:17]: what is `FOO` here and above? I'm not sure how to read this...

File: test/lit/wasm-split/call_exports.mjs (hunk 2 above)
Comment (tlively) [reply test/lit/wasm-split/call_exports.mjs:6]: This is not a lit test, this is a script invoked by profile-guided.wast (which _is_ a lit test).

File: src/tools/wasm-split.cpp (hunk 1 above)
Comment (tlively) [reply src/tools/wasm-split.cpp:449]: For now the only thing to compare against is the code that writes the profile, which is also in this file.

File: test/lit/wasm-split/profile-guided.wast (hunk 3 above)
Comment (tlively) [reply test/lit/wasm-split/profile-guided.wast:3]: That's correct, but it also sets up state that is used in the subsequent `RUN` lines. Specifically, it writes the file %t.instrumented.wasm. The "%t" is replaced with a unique filename at runtime.

File: test/lit/wasm-split/profile-guided.wast (hunk 4 above)
Comment (tlively) [reply test/lit/wasm-split/profile-guided.wast:17]: `FOO` takes the place of `CHECK` (via the --check-prefix option above). This option allows multiple different test outputs to be checked in the same file with each using a different check prefix.

Comment (aheejin) [reply test/lit/wasm-split/profile-guided.wast:17]: Reference: https://llvm.org/docs/CommandGuide/FileCheck.html#the-filecheck-check-prefix-option

File: src/tools/wasm-split.cpp (hunk 1 above)
Comment (aheejin) [reply src/tools/wasm-split.cpp:449]: Maybe it'd be good to add comments on the file format, however simple it is.

**Review by aheejin — APPROVED** [review]

File: test/lit/wasm-split/profile-guided.wast (hunk 4 above)
Comment (kripken) [reply test/lit/wasm-split/profile-guided.wast:17]: Thanks! I see, so the relevant docs are for FileCheck and not lit... TIL

I understand the benefit of the prefix in the example there, but not here. It seems like FOO, BAR, are each used once and used immediately. Could they all have been CHECK?

Comment (tlively) [reply test/lit/wasm-split/profile-guided.wast:17]: If the had all been CHECK, then every invocation of filecheck would have checked all of them against the input, which wouldn't have worked. Remember that filecheck matches against the entire file, not just the lines after or between RUN lines. filecheck doesn't know anything about RUN lines at all.

Comment (kripken) [reply test/lit/wasm-split/profile-guided.wast:17]: Thanks, I had assumed it did understand about RUN lines from the examples, but now I get it..

**Pull request merged**